### PR TITLE
bootloader: b0: Add a minimal overlay config

### PR DESCRIPTION
**Pull request by kyberdin**
Adds an overlay config for building a minimal b0 image.
Moves commented out config changes for smaller image size from
prj.conf to the overlay.

Ref: NCSDK-6836
Signed-off-by: Stephen Stauts <stephen.stauts@nordicsemi.no>

### CODE REVIEW

**Comment by hakonfam** [comment]
Draft PR, waiting for finishing touches before review begins.

File: samples/bootloader/overlay-minimal-size.conf
Comment (hakonfam) [inline samples/bootloader/overlay-minimal-size.conf:1]: IMO it would be good to have a pointer to this file and a brief explanation on how to include it. There should be some examples similar to this already (just grep for 'minimal' in the .rst docs).

Comment (kyberdin) [reply samples/bootloader/overlay-minimal-size.conf:1]: I've already added such documentation into the `hello_world` sample that will be in a separate PR.

Perhaps it would be simpler to just close this PR and move this addition to that one when it's ready?

Comment (hakonfam) [reply samples/bootloader/overlay-minimal-size.conf:1]: Got it. I think that information should be available in the bootloader sample documentation as well, since users might not use the hello world sample for information.

Its up to you whether you choose to include it in this PR or another, I have no preference on that.

Comment (kyberdin) [reply samples/bootloader/overlay-minimal-size.conf:1]: You're right that the information is probably more relevant in this README than the sample. Moving it over here and into this PR works for me :+1: might make the other sample's documentation a bit more concise as well.

Comment (kyberdin) [reply samples/bootloader/overlay-minimal-size.conf:1]: Added in latest commit. Also opened the PR to be reviewed, since this the majority of the task's content.

**Comment by kyberdin** [comment]
Refactored the last commit to acommodate the associated task NCSDK-7365, adding more information about the child-image overlays and config options to the Multi-Image documentation.

**Review by oyvindronningstad — APPROVED** [review]
Looks very good!

**Review by FrancescoSer — APPROVED** [review]
Approving, considering that soon the included RST files will be modified again.

**Comment by kyberdin** [comment]
@FrancescoSer I had to tweak some formatting in the UICR note at `samples/bootloader/README.rst:90` to make CI happy. Would you mind reviewing the changes and re-approving if they're OK?

**Review by FrancescoSer — APPROVED** [review]

File: doc/nrf/ug_multi_image.rst
Comment (tejlmand) [inline doc/nrf/ug_multi_image.rst:169]: How does a user know where to find this list: `SHARED_MULTI_IMAGE_VARIABLES`.

We should not refer to such a list without providing more details.
I think that it should be enough to state:
```suggestion
* CMake build settings, such as `BOARD_DIR`, build type, Toolchain info, Partition manager info, and similar are always passed to child images.
```
@FrancescoSer  please review the text suggestion.

File: doc/nrf/ug_multi_image.rst
Comment (tejlmand) [inline doc/nrf/ug_multi_image.rst:168]: The `\_` is shown when you use: \`\`\\_\`\`
```suggestion
* Any variable named ``<IMAGE_NAME>_FOO`` in a parent image is propagated to the child image named ``<IMAGE_NAME>`` as ``FOO``.
```

File: doc/nrf/ug_multi_image.rst
Comment (tejlmand) [inline doc/nrf/ug_multi_image.rst:203]: Overlays are used by dts.
This is a Kconfig fragment.

Also, the `prj.conf` is just the samples `prj.conf`, and a custom conf file can be name anything.
Doesn't have to be named `prj.conf`.

> To always build the application with these configurations

Which configurations ?
Placing `mcuboot_CONFIG_LOG_MINIMAL=y` in my applications `prj.conf` is not going to work.
That setting would have to go into the config file for `mcuboot`.
That is not clear from this text.

File: doc/nrf/ug_multi_image.rst
Comment (tejlmand) [inline doc/nrf/ug_multi_image.rst:192]: This becomes a very long rendered line.

Not sure that is really helpful.

File: doc/nrf/ug_multi_image.rst
Comment (tejlmand) [inline doc/nrf/ug_multi_image.rst:260]: Why not do this the proper CMake way ?

```suggestion
        west build -b <board_name> -- -DCMAKE_VERBOSE_MAKEFILE=1 -DCMAKE_BUILD_PARALLEL_LEVEL=1
```

Then you can remove the distinction between Linux/MacOS/Windows, as they all work the same way.

File: samples/bootloader/README.rst
Comment (tejlmand) [inline samples/bootloader/README.rst:189]: Kconfig fragments, not overlays.

**Review by tejlmand — CHANGES_REQUESTED** [review]

File: doc/nrf/ug_multi_image.rst
Comment (kyberdin) [reply doc/nrf/ug_multi_image.rst:192]: This is changed in #3607 to be more readable. A lot of the docs-specific changes were not considered here due to them changing again in that PR, as mentioned [here](https://github.com/nrfconnect/sdk-nrf/pull/3518#pullrequestreview-559201567).

**Comment by kyberdin** [comment]
@tejlmand The documentation suggestions that you made here are good ones and should be addressed. However, would it be acceptable with you to address them in #3607, where the docs change into a more modern state? @FrancescoSer and I already agreed to overlook more detailed doc review here in favor of the commits in that PR, which builds on top of the two commits introduced by this one.

File: doc/nrf/ug_multi_image.rst
Comment (tejlmand) [inline doc/nrf/ug_multi_image.rst:200]: This whole section does not belong in the multi image guide, most of it is already described here:
http://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/ug_bootloader.html#adding-a-bootloader-chain-to-your-application

and the additions in this section belongs there as well.

(Seems I forgot to hit `add comment` on this comment)

**Review by tejlmand — CHANGES_REQUESTED** [review]
> @tejlmand The documentation suggestions that you made here are good ones and should be addressed. However, would it be acceptable with you to address them in #3607, where the docs change into a more modern state? @FrancescoSer and I already agreed to overlook more detailed doc review here in favor of the commits in that PR, which builds on top of the two commits introduced by this one.

Seems one of my comments (the most important one actually), was missed in my review.
All the bootloader stuff should go out of `ug_multi_image.rst` before merging.

The rest can be addressed in #3607.

File: doc/nrf/ug_multi_image.rst
Comment (kyberdin) [reply doc/nrf/ug_multi_image.rst:200]: I would agree if the intent of the lines here was also to be a guide of how to build with the bootloader. However, the intent of these lines is for clearly explaining what each configuration option does in relation to their specific image in the build; which is ultimately the purpose of this section of the multi-image docs.

Removing the lines from this PR is perfectly fine with me. I still support them being a part of this page in #3607 though since they add clarity through example.

Of course, we can also change the example to not be so much of a duplicate, but the immutable/upgradable integration is a good example for the concepts of this section.

File: doc/nrf/ug_multi_image.rst
Comment (kyberdin) [reply doc/nrf/ug_multi_image.rst:168]: Addressed in [this commit](https://github.com/nrfconnect/sdk-nrf/pull/3607/commits/8f78f39ea2d00a43e1149293706416747f825acd) for #3607.

File: doc/nrf/ug_multi_image.rst
Comment (kyberdin) [reply doc/nrf/ug_multi_image.rst:203]: Addressed in [this commit](https://github.com/nrfconnect/sdk-nrf/pull/3607/commits/8f78f39ea2d00a43e1149293706416747f825acd) for #3607.

File: doc/nrf/ug_multi_image.rst
Comment (kyberdin) [reply doc/nrf/ug_multi_image.rst:260]: Addressed in [this commit](https://github.com/nrfconnect/sdk-nrf/pull/3607/commits/8f78f39ea2d00a43e1149293706416747f825acd) for #3607.

File: samples/bootloader/README.rst
Comment (kyberdin) [reply samples/bootloader/README.rst:189]: Addressed in [this commit](https://github.com/nrfconnect/sdk-nrf/pull/3607/commits/8f78f39ea2d00a43e1149293706416747f825acd) for #3607.

Also updated wording from overlay to "fragment" or "Kconfig fragment" in other documents being updated in #3607.

File: doc/nrf/ug_multi_image.rst
Comment (tejlmand) [reply doc/nrf/ug_multi_image.rst:200]: No, this doesn't add clarity, having all of those in a single line makes it less generic.

What we could write would be something like:

> It is possible to provide configuration settings for child images, either as individual settings or through Kconfig fragments.
> Each child image is referenced using its image name, so to set the configuration option `CONFIG_FOO=val` in the child image `bar`, you can use:
> ```
> cmake -Dbar_CONFIG_FOO=val ...
> ```
>
> A Kconfig fragment can be added to the child image default configuration in similar way.
> To add an extra Kconfig fragment `baz.conf` to `bar`, you can use
> ```
> cmake -Dbar_OVERLAY_CONFIG=baz.conf ...
> ```
>
> It is also possible to provide a custom configuration file as replacement for the default Kconfig file for the child image.
> To use the  custom configuration file ``myfile.conf`` when building `bar`, you can use
> ```
> cmake -Dbar_CONF_FILE=myfile.conf ...
> ```
>
> If your application is including multiple child image, then you can combine the above together, that is:
> - Setting `CONFIG_FOO=val` in main application
> - Adding a Kconfig fragment `baz.conf` to `bar` child image, using `-Dbar_OVERLAY_CONFIG=baz.conf`
> - Using `myfile.conf` as configuration for the `quz` child image, using `-Dquz_CONF_FILE=myfile.conf`
>
> will become:
> ```
> cmake -DCONFIG_FOO=val -Dbar_OVERLAY_CONFIG=baz.conf -Dquz_CONF_FILE=myfile.conf ....
> ```

Comment (kyberdin) [reply doc/nrf/ug_multi_image.rst:200]: @FrancescoSer thoughts about this?

**Comment by kyberdin** [comment]
@oyvindronningstad @tejlmand After a chat with @hakonfam it was decided to change this implementation from `overlay-minimal-size.conf` usage to `prj_minimal.conf` with `board/xxx_minimal.conf` fragments so that the minimal b0 image can suit the board that it is being compiled for.

For example, right now building b0 will result in a 14kB image that using a 32kB partition size. This is wasteful, especially for devices like the nRF52811 in which +16kB app space would be very nice to have. Additionally, the presence of a provision image means the `PM_PARTITION_SIZE_B0_IMAGE` value will differ between boards with OTP and those without.

File: doc/nrf/ug_multi_image.rst
Comment (FrancescoSer) [inline doc/nrf/ug_multi_image.rst:188]: @tejlmand @kyberdin I just added the section about variables in child images, as mentioned in one of the previous comments.

**Review by FrancescoSer — APPROVED** [review]

**Review by tejlmand — CHANGES_REQUESTED** [review]
@kyberdin please cleanup the commits, and remove the merge commit, so it is easier to review the final result.

**Comment by kyberdin** [comment]
The commits have been squashed and now represent its final form.

File: subsys/bootloader/Kconfig
Comment (oyvindronningstad) [inline subsys/bootloader/Kconfig:173]: 0x2e00 + 0x1000 = 0x3e00

What about the last 0x200 to get to 0x4000 (16k)?

Comment (kyberdin) [reply subsys/bootloader/Kconfig:173]: I took the `0x3e00` from [sdk-mcuboot's external crypto overlay](https://github.com/nrfconnect/sdk-mcuboot/blob/master/boot/zephyr/overlay-minimal-external-crypto.conf#L46). The `0x200` is MCUboot's pad size, but since that isn't relevant here I saw this difference as forcing it to compile less than 16kB while keeping a consistent value across bootloaders.

Using `0x4000` is fine with me though, as it's more relevant to b0.

Comment (kyberdin) [reply subsys/bootloader/Kconfig]: Using 0x3000 instead of 0x2e00 makes much more sense. Updated the value.

File: subsys/bootloader/Kconfig
Comment (tejlmand) [inline subsys/bootloader/Kconfig:173]: This commit should be squashed in here:
0899f85d792082ddac0b83f956d34762f243bfaa

Please avoid changing the same line in multiple commit in the same PR, as it make it harder to review the PR, commit-by-commit.

**Review by tejlmand — CHANGES_REQUESTED** [review]
Looks good.
Only need cleanup of commits.

File: subsys/bootloader/Kconfig
Comment (kyberdin) [reply subsys/bootloader/Kconfig:173]: Sorry about that. Planned on cleaning up the commits after approval, but I'll just force push the rebased change(s) in the future to keep the commit history clean 👍

**Comment by tejlmand** [comment]
@kyberdin please rebase, and fix conflicts.

**Pull request merged**
